qiskit_qec/linear/matrix.py

```diff
@@ -493,3 +493,281 @@ def istack(mat: np.ndarray, size: int, interleave: bool = False) -> np.ndarray:
+    """Performs span-preserving row operations on a matrix in the ring Z/NZ. These include:
+
+    1. Swap two rows mat[i] and mat[j];
+    2. Multiply a row mat[i] by a scalar c (valid only when c is a unit), i.e., mat[i] = c * mat[i];
+    3. Add a multiple of one row to another, i.e., mat[i] = mat[i] + c * mat[j];
+    4. Append the product of a row by a scalar c to the end of matrix (used when c is a zero divisor), i.e., mat = mat.append(c * mat[i]);
+    5. Update two rows by multiplying by a full-rank 2x2 matrix, i.e., mat[i] = a * mat[i] + b * mat[j], mat[j] = c * mat[i] + d * mat[j].
+    
+    Args:
+        mat: input matrix
+        row_op: tuple (op, rows, coeff), where op is the operation to be performed, rows are the row indices, and coeff is a list of coefficients for the operation (empty list if op is 'swap')
+        N: modulus
+
+    Returns:
+        matrix after performing the row operation
+    """
+    op, rows, coeff = row_op
+    if op == 'swap':
+        mat[rows[0]], mat[rows[1]] = mat[rows[1]], mat[rows[0]]
+    elif op == 'unit':
+        mat[rows[0]] = np.mod(coeff[0] * mat[rows[0]], N)
+    elif op == 'add':
+        mat[rows[0]] = np.mod(mat[rows[0]] + coeff[0] * mat[rows[1]], N)
+    elif op == 'append':
+        mat = np.vstack((mat, np.mod(coeff[0] * mat[rows[0]], N)))
+    elif op == 'update':
+        R1 = np.mod(coeff[0] * mat[rows[0]] + coeff[1] * mat[rows[1]], N)
+        R2 = np.mod(coeff[2] * mat[rows[0]] + coeff[3] * mat[rows[1]], N)
+        mat[rows[0]], mat[rows[1]] = R1, R2
+
+    return mat
+    
+
+def howell(mat: np.ndarray, N: int) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+    """Computes the Howell form of a matrix in the ring Z/NZ, the corresponding transformation matrix, and the kernel.
+
+    Args:
+        mat: input matrix
+        N: modulus
+
+    Returns:
+        H: Howell form of mat
+        U: transformation matrix (U @ mat = H)
+        K: kernel of mat (mat @ K = 0)
+
+    Examples:
+        >>> mat = numpy.array([[8, 5, 5],
+                               [0, 9, 8],
+                               [0, 0, 10]])
+        >>> N = 12
+        >>> H, U, K = howell(mat, N)
+        >>> H
+        array([[4, 1, 0],
+               [0, 3, 0],
+               [0, 0, 1]])
+        >>> U
+        array([[8, 1, 0],
+               [0, 7, 4],
+               [9, 3, 4]])
+        >>> K
+        array([[6, 6, 6],
+               [0, 4, 4]])
+    """
+    H = mat.copy()
+    U = np.eye(H.shape[0], dtype=int)
+    m, n = H.shape
+    row_ops = []
+    
+    r = 0
+    # going through each column
+    for c in range(n):
+        # find j such that H[j, c] > 0
+        j = r
+        while j < m and H[j, c] == 0:
+            j += 1
+        if j < m:
+            # found j: if j > r, swap rows r and j
+            if j > r:
+                row_ops.append(('swap', [r, j], []))
+                H = _do_row_op(H, row_ops[-1], N)
```

Comment:
This approach is rather expensive and causes a branch each time _do_row_op is called. It would be better and faster to create different row op methods for each time and avoid the conditional branch. Here you know what you want to do and append the op to row_ops but then immediately recover the result twice to but into _do_row_op which then does a conditional branch ...

So if you want to do a `unit` then simply call

_do_unit_op(H,N)

Also we should look at in-place and not-in_place options as there could be a lot of coping and reassigning which would again slow down the algorithm. I am find having two methods or one method with an option. Probably best to have a higher level option with both methods and then two separate methods for in-place and not-in_place methods.

Reply:
Hi @dsvandet, I have two questions regarding this comment:

1. For each individual row operation, do we also want to create a low-level and a high-level method at the same time, e.g., `do_unit_op` and `_do_unit_op`? Since these row operations are just being used internally for `howell`, I'm not sure if it's necessary.

2. I don't fully understand what you meant by in-place vs. not-in_place options. In the current implementation, we are reassigning values to `x` and `b` multiple times inside the loop. So with the not-in_place option, are you suggesting not do that? What would be an alternative way? Could you please maybe provide an example?

Thanks!

Reply:
HI @ruihao-li,

1. I would just implement internal `_do_init_op` etc. but have a general "do_row_op" method that has all the options etc that calls the specific methods.

2. By in-place I was meaning that the primary object, H say, would be modified directly with creating copies and then realigning back. However, I had another look at the code and you are already passing by reference and doing in-place assignments and returning the reference so I miss-wrote.